app/src/org/commcare/tasks/DataPullTask.java

```diff
@@ -102,258 +101,296 @@ public DataPullTask(String username, String password,
-                    } else if (returnCode == PROGRESS_RECOVERY_FAIL_BAD) {
-                        //WELL! That wasn't so good. TODO: Is there anything 
-                        //we can do about this?
-
-                        //wipe our login if one happened
-                        if (loginNeeded) {
-                            CommCareApplication._().releaseUserResourcesAndServices();
-                        }
-                        this.publishProgress(PROGRESS_DONE);
-                        return new ResultAndError<>(PullTaskResult.UNKNOWN_FAILURE, failureReason);
-                    }
-
-                    if (loginNeeded) {
-                        CommCareApplication._().releaseUserResourcesAndServices();
-                    }
-                } else if (pullResponse.responseCode == 500) {
-                    if (loginNeeded) {
-                        CommCareApplication._().releaseUserResourcesAndServices();
-                    }
-                    Logger.log(AndroidLogger.TYPE_USER, "500 Server Error|" + username);
-                    return new ResultAndError<>(PullTaskResult.SERVER_ERROR, "");
-                }
-            } catch (SocketTimeoutException e) {
-                e.printStackTrace();
-                Logger.log(AndroidLogger.TYPE_WARNING_NETWORK, "Timed out listening to receive data during sync");
-                responseError = PullTaskResult.CONNECTION_TIMEOUT;
-            } catch (ConnectTimeoutException e) {
-                e.printStackTrace();
-                Logger.log(AndroidLogger.TYPE_WARNING_NETWORK, "Timed out listening to receive data during sync");
-                responseError = PullTaskResult.CONNECTION_TIMEOUT;
-            } catch (ClientProtocolException e) {
-                e.printStackTrace();
-                Logger.log(AndroidLogger.TYPE_WARNING_NETWORK, "Couldn't sync due network error|" + e.getMessage());
-            } catch (UnknownHostException e) {
-                Logger.log(AndroidLogger.TYPE_WARNING_NETWORK, "Couldn't sync due to bad network");
-                responseError = PullTaskResult.UNREACHABLE_HOST;
-            } catch (IOException e) {
-                // TODO Auto-generated catch block
-                e.printStackTrace();
-                Logger.log(AndroidLogger.TYPE_WARNING_NETWORK, "Couldn't sync due to IO Error|" + e.getMessage());
             }
-            if (loginNeeded) {
-                CommCareApplication._().releaseUserResourcesAndServices();
+        };
+    }
+
+    private byte[] getEncryptionKey() {
+        byte[] key;
+        if (loginNeeded) {
+            initUKRForLogin();
+            if (ukrForLogin == null) {
+                return null;
+            }
+            key = ByteEncrypter.wrapByteArrayWithString(ukrForLogin.getEncryptedKey(), password);
+        } else {
+            //Only purge cases if we already had a logged in user. Otherwise we probably can't read the DB.
+            CaseUtils.purgeCases();

```

Comment:
I don't think this is the best place for this call now. We need this to happen before we perform a sync request for sure, but if we've refactored this modularly in this process, I think we should call this after getEncryptionKey() has returned with a key outside of this method.